### ISSUE TITLE
Incorrect type error for wrapped defining use of TAIT/ATPIT

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-analyzer#15704]
```rust
#![feature(impl_trait_in_assoc_type)]

trait WrappedAssoc {
    type Assoc;
    fn do_thing(&self) -> Option<Self::Assoc>;
}

struct Foo;
impl WrappedAssoc for Foo {
    type Assoc = impl core::fmt::Debug;

    fn do_thing(&self) -> Option<Self::Assoc> {
        Some(())
    }
}
```

rust-analyzer incorrectly reports:

> expected Option&lt;{unknown}>, found Option&lt;()>

* No error unless the associated type is inside a generic type, i.e. `-> Self::Assoc {` does not error.
* No error in a non-trait function.
* Does error if the type is declared as `type_alias_impl_trait`, as long as the defining occurrence is still by way of the associated type.

    ```rust
    type Opaque = impl core::fmt::Debug;
    ... 
        type Assoc = Opaque;
    ```

Of course, these are unstable language features, but since there is desire to stabilize them soon, I figured this report would be of interest.

**rust-analyzer version**: 0.3.1681-standalone (0840038f0 2023-09-30)

**rustc version**: rustc 1.74.0-nightly (8ce4540bd 2023-09-29)

**Comment by traviscross** [comment]
Here's an interesting caveat:

```rust
#![feature(impl_trait_in_assoc_type)]

// Interestingly, if we comment out this line, the problem goes away.
#[cfg(False)]
pub enum Option<T> {
    None,
    Some(T),
}

trait Trait {
    type Assoc;
    const DEFINE: Option<Self::Assoc>;
}

impl Trait for () {
    type Assoc = impl Sized;
    const DEFINE: Option<Self::Assoc> = Option::Some(());
    //~^ r-a ERROR expected Option<{unknown}>, found Option<()> [E0308]
    //~| ...but no error if `Option` is defined above.
}
```

The problem goes away if `Option` is defined locally here.

**Comment by Veykril** [comment]
That is not the case for me 😅 I get the error no matter what (which is what I'd expect)
